src/pickup.cpp

```diff
@@ -72,52 +73,134 @@ struct pickup_count {
+            return rule_state::WHITELISTED;
+        }
+    } else {
+        return rule_state::BLACKLISTED;
+    }
+    return rule_state::NONE;
+}
+
+static void process_whitelisted_autopick( item *what, tripoint where )
+{
+    if( !what->is_container() ) {
+        return;
+    }
+    for( item *entry : what->all_items_top() ) {
+        if( should_auto_pickup( entry ) == rule_state::BLACKLISTED ) {
+            // blacklisted items should be removed from the container
+            // and dropped on the same tile the container is on
+            get_map().add_item( where, what->remove_item( *entry ) );
+        }
+    }
+}
+
+static std::vector<item_location> get_pickup_list_from( item_location &container )
+{
+    item *container_item = container.get_item();
+    bool pick_all_items = true;
+
+    std::vector<item_location> pickup_list;
+    std::list<item *> contents = container_item->all_items_top();
+    pickup_list.reserve( contents.size() );
+
+    for( item *item_to_check : contents ) {
+        const rule_state pickup_state = should_auto_pickup( item_to_check );
+        if( pickup_state == rule_state::WHITELISTED ) {
+            // whitelisted containers should exclude contained blacklisted items
+            process_whitelisted_autopick( item_to_check, container.position() );
+            // pick up the whitelisted item
+            pickup_list.emplace_back( container, item_to_check );
+        } else if( pickup_state == rule_state::BLACKLISTED || item_to_check->is_container_empty() ) {
+            // skip empty containers and blacklisted items
+            pick_all_items = false;
+        } else {
+            // get pickup list from nested item container
+            item_location location = item_location( container, item_to_check );
+            std::vector<item_location> pickup_nested = get_pickup_list_from( location );
+
+            // container with content was NOT marked for pickup
+            if( pickup_nested.size() != 1 || pickup_nested[0] != location ) {
+                pick_all_items = false;
+            }
+            pickup_list.reserve( pickup_nested.size() + pickup_list.size() );
+            pickup_list.insert( pickup_list.end(), pickup_nested.begin(), pickup_nested.end() );
+        }
+    }
+    // all items in container were approved for pickup
+    if( !contents.empty() && pick_all_items ) {
+        bool all_batteries = true;
+        bool powered_container = container_item->ammo_capacity( ammo_battery );
+        if( powered_container ) {
+            // when dealing with battery powered tools there should only be one pocket
+            // and one battery inside but this could change in future so account for that here
+            for( size_t i = 0; i < pickup_list.size() && all_batteries; i++ ) {
+                item *entry = pickup_list[i].get_item();
+                if( !entry->is_battery() ) {
+                    all_batteries = false;
                 }
             }
```

Comment:
This can be simplified to
```cpp
all_batteries = std::all_of( pickup_list.begin(), pickup_list.end(),
                             []( const item_location &il ) {
                                 return il.get_item()->is_battery();
                             } );
```

Reply:
Resolved in commit 7b670045403370fc9b7132711d2dc715a1422270.